Fix part of #10798: Fix froms.js file

suryasiriki4 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #10798.
2. This PR does the following: fixes e2e tests with action.js and waitFor.js.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

suryasiriki4 committed:
fix froms.js

suryasiriki4 committed:
added test before button names

suryasiriki4 commented:
@U8NWXD , @kevintab95  PTAL

suryasiriki4 committed:
Update forms.js

U8NWXD commented:
@suryasiriki4 a lot of the e2e tests are failing, which doesn't usually happen. Could you look into the logs and see if you can figure out what's going wrong? I suspect there's a bug in your PR

suryasiriki4 committed:
imported action from action.js

suryasiriki4 committed:
imported action from action.js

suryasiriki4 committed:
removed `.sendKeys()` in _appendContextText

also changed names of variables in AutocompleteDropdownEditor

suryasiriki4 commented:
> @suryasiriki4 a lot of the e2e tests are failing, which doesn't usually happen. Could you look into the logs and see if you can figure out what's going wrong? I suspect there's a bug in your PR

@U8NWXD , I have resolved the bug. PTAL.

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -51,7 +52,7 @@ var GraphEditor = function(graphInputContainer) {
+    await action.click('Test Add Node Button', addNodeButton);
     // Offsetting from the graph container.
     await browser.actions().mouseMove(
       graphInputContainer, {x: xOffset, y: yOffset}).perform();
```
Can we make sure that the graphInputContainer is visible first?

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -61,7 +62,7 @@ var GraphEditor = function(graphInputContainer) {
-    await addEdgeButton.click();
+    await action.click('Test Add Edge Button', addEdgeButton);
     await browser.actions().mouseMove(
       vertexElement(vertexIndex1)).perform();
```
ditto with the `vertexElement(vertexIndex1)` and `vertexElement(vertexIndex2)`

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -51,7 +52,7 @@ var GraphEditor = function(graphInputContainer) {
     var addNodeButton = graphInputContainer.element(
       by.css('.protractor-test-Add-Node-button'));
-    await addNodeButton.click();
+    await action.click('Test Add Node Button', addNodeButton);
```
Here and elsewhere, please remove the `Test ` at the start of the element names

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -89,10 +90,10 @@ var GraphEditor = function(graphInputContainer) {
+        await action.click(`Vertex Element ${i}`, vertexElement(i));
       }
     },
     expectCurrentGraphToBe: async function(graphDict) {
```
Below, please add a waitFor before the `.count()` call. The waitFor should only execute when `edgesList.length > 0`

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -170,31 +174,34 @@ var ListEditor = function(elem) {
-    await elem.element(by.css('.' + buttonName)).click();
+    var clickToolbarButton = elem.element(by.css('.' + buttonName));
+    await action.click(
+      `Click Toolbar Button: ${buttonName}`, clickToolbarButton);
```
```suggestion
      `Toolbar Button: ${buttonName}`, clickToolbarButton);
```

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -170,31 +174,34 @@ var ListEditor = function(elem) {
       await (await elem.all(by.css('.oppia-rte')).first()).isPresent()
     ).toBe(true);
-    await (await elem.all(by.css('.oppia-rte')).first()).clear();
+    var firstOfOppiaRte = await elem.all(by.css('.oppia-rte')).first();
+    await action.clear('First of oppia rte', firstOfOppiaRte);
```
Here and elsewhere, please make the RTE element names clearer. For example:

```suggestion
    var firstOppiaRte = await elem.all(by.css('.oppia-rte')).first();
    await action.clear('First Rich Text Editor', firstOppiaRte);
```

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -294,24 +301,29 @@ var SetOfTranslatableHtmlContentIdsEditor = function(elem) {
+      await action.click('Select2 Container', select2Container);
       var actualOptions = await element(by.css('.select2-dropdown'))
         .all(by.tagName('li')).map(
           async function(optionElem) {
```
Below, please add a waitFor before the .getText. You can also you our new `action.getText` function!

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -340,13 +355,15 @@ var AutocompleteMultiDropdownEditor = function(elem) {
+          'Select2 Search Field', select2SearchField, texts[i] + '\n');
       }
     },
     expectCurrentSelectionToBe: async function(expectedCurrentSelection) {
```
ditto: getText

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -407,8 +428,10 @@ var MultiSelectEditor = function(elem) {
+        'Test Search Bar Dropdown Toggle', testSearchBarDropdownToggle);
 
       // Find the selected elements.
       var actualSelection = await elem.element(
```
ditto: getText

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -420,8 +443,10 @@ var MultiSelectEditor = function(elem) {
+        'Test Search Bar Dropdown Toggle', testSearchBarDropdownToggle);
     }
   };
 };
```
ditto: the many getText calls below. Please handle the getTagName and getAttribute calls below similarly

suryasiriki4 replied:
done!

U8NWXD reviewed (changes requested):
Thanks for your PR @suryasiriki4! I left some comments on how you can make it even better

suryasiriki4 committed:
Update core/tests/protractor_utils/forms.js

removed Click before Toolbar Button in _clickToolbarButton routine

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

suryasiriki4 committed:
Remved `test` at the start of Element names

suryasiriki4 committed:
Ensured Visibility of some elements

ensured visibility of graphInputContainer, vertexElement(vertexIndex1), vertexElement(vertexIndex2).

suryasiriki4 committed:
ensured visiblity of allEdgesElement

suryasiriki4 committed:
Made Rte element names clearer

suryasiriki4 committed:
replaced elem.getText() with action.getText()

suryasiriki4 committed:
removed extra tabs in the front

suryasiriki4 committed:
removed unnecessary indentation

suryasiriki4 committed:
removed extra indentation in `RichTextChecker`

suryasiriki4 committed:
Added waitFor before .getTagName(), .getAttribute

suryasiriki4 commented:
@U8NWXD , I have made all the changes recommended by you. PTAL.

suryasiriki4 committed:
remove index number in error message

suryasiriki4 committed:
replaced backticks with apostrophes

suryasiriki4 committed:
Merge remote-tracking branch 'upstream/develop' into fix-forms

suryasiriki4 committed:
resolved merege conflicts locally

suryasiriki4 committed:
Merge remote-tracking branch 'origin/fix-forms' into fix-forms

suryasiriki4 commented:
@U8NWXD, PTAL

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -61,10 +64,16 @@ var GraphEditor = function(graphInputContainer) {
       by.css('.protractor-test-Add-Edge-button'));
-    await addEdgeButton.click();
+    await action.click('Add Edge Button', addEdgeButton);
+    await waitFor.visibilityOf(
+      vertexElement(
+        vertexIndex1), `Vertex element ${vertexIndex1} taking too long`);
```
I think this indentation is easier to read:

```suggestion
    await waitFor.visibilityOf(
      vertexElement(vertexIndex1),
      `Vertex element ${vertexIndex1} taking too long`);
```

Also, `taking too long` is a little vague. Taking too long to do what?

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -61,10 +64,16 @@ var GraphEditor = function(graphInputContainer) {
     await browser.actions().mouseDown().perform();
+    await waitFor.visibilityOf(
+      vertexElement(
+        vertexIndex2), `Vertex element ${vertexIndex2} taking too long`);
```
ditto

suryasiriki4 replied:
done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -407,21 +442,26 @@ var MultiSelectEditor = function(elem) {
         .map(async function(selectedElem) {
-          return await selectedElem.getText();
+          return await action.getText(
+            `Search bar dropdown menu ${selectedElem}`, selectedElem);
```
What kind of string does this produce? I don't know how elements get stringified. If the string isn't helpful, you could use a for loop instead to keep track of the index

suryasiriki4 replied:
I have renamed it as 'selected element', I think it should be sufficient. done!

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -497,6 +538,9 @@ var RichTextChecker = async function(arrayOfElems, arrayOfTexts, fullText) {
 
   var _readFormattedText = async function(text, tagName) {
+    await waitFor.visibilityOf(
+      await arrayOfElems.get(
+        arrayPointer), `element at ${arrayPointer} is not visible`);
```
I think this line break pattern is more readable

```suggestion
      await arrayOfElems.get(arrayPointer),
      `element at ${arrayPointer} is not visible`);
```

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -537,7 +583,10 @@ var RichTextChecker = async function(arrayOfElems, arrayOfTexts, fullText) {
         args.push(arguments[i]);
       }
-      expect(await elem.getText()).toBe(arrayOfTexts[arrayPointer]);
+      expect(
+        await action.getText(
+          `${elem} in Array of Elements in Rich Text Area`, elem))
+        .toBe(arrayOfTexts[arrayPointer]);
```
The current indentation makes it hard to keep track of which parentheses go together. This might be clearer:

```suggestion
      expect(
        await action.getText(
          `${elem} in Array of Elements in Rich Text Area`, elem)
      ).toBe(arrayOfTexts[arrayPointer]);
```

U8NWXD commented on core/tests/protractor_utils/forms.js:
```diff
@@ -618,6 +667,9 @@ var CodeMirrorChecker = function(elem, codeMirrorPaneToScroll) {
       await browser.executeScript(
         '$(\'.CodeMirror-vscrollbar\').' + codeMirrorPaneToScroll +
         '().scrollTop(' + String(scrollTo) + ');');
+      await waitFor.visibilityOf(
+        elem.element(by.css(
+          '.CodeMirror-linenumber')), 'code mirror linenumber is not visible');
```
ditto: clearer indentation

suryasiriki4 replied:
done!

U8NWXD reviewed (changes requested):
This is getting a lot better @suryasiriki4! I left a few more comments

suryasiriki4 committed:
Update core/tests/protractor_utils/forms.js

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

suryasiriki4 committed:
Update core/tests/protractor_utils/forms.js

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

suryasiriki4 committed:
Update core/tests/protractor_utils/forms.js

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

suryasiriki4 committed:
Merge branch 'oppia:develop' into fix-forms

suryasiriki4 committed:
removed the element names in strings

made all the changes recommended in the second review

suryasiriki4 committed:
breaked the line after '('

suryasiriki4 commented:
> This is getting a lot better @suryasiriki4! I left a few more comments

@U8NWXD, I have made the changes recommended. PTAL.

U8NWXD reviewed (approved):
LGTM. Thanks @suryasiriki4!

U8NWXD commented:
@suryasiriki4 I'm seeing an unfamiliar error message:

```
1) Exploration history should display the history
  Message:
    Failed: Line number element is not visible for getText()
    Wait timed out after 10068ms
  Stack:
    TimeoutError: Line number element is not visible for getText()
    Wait timed out after 10068ms
        at /home/runner/work/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:2201:17
        at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
    From asynchronous test: 
    Error
        at Suite.<anonymous> (/home/runner/work/oppia/oppia/core/tests/protractor_desktop/explorationHistoryTab.js:67:3)
        at Object.<anonymous> (/home/runner/work/oppia/oppia/core/tests/protractor_desktop/explorationHistoryTab.js:31:1)
        at Module._compile (internal/modules/cjs/loader.js:1063:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
        at Module.load (internal/modules/cjs/loader.js:928:32)
        at Function.Module._load (internal/modules/cjs/loader.js:769:14)
```

It doesn't look like one of the flakes we've seen before. Could you dig into it a little? Here's the build link: https://github.com/oppia/oppia/pull/12699/checks?check_run_id=2525204687. If you need help debugging, check out the E2E test wiki page at https://github.com/oppia/oppia/wiki/End-to-End-Tests

suryasiriki4 committed:
Merge branch 'oppia:develop' into fix-forms

suryasiriki4 commented:
> @suryasiriki4 I'm seeing an unfamiliar error message:
> 
> ```
> 1) Exploration history should display the history
>   Message:
>     Failed: Line number element is not visible for getText()
>     Wait timed out after 10068ms
>   Stack:
>     TimeoutError: Line number element is not visible for getText()
>     Wait timed out after 10068ms
>         at /home/runner/work/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:2201:17
>         at runMicrotasks (<anonymous>)
>         at processTicksAndRejections (internal/process/task_queues.js:93:5)
>     From asynchronous test: 
>     Error
>         at Suite.<anonymous> (/home/runner/work/oppia/oppia/core/tests/protractor_desktop/explorationHistoryTab.js:67:3)
>         at Object.<anonymous> (/home/runner/work/oppia/oppia/core/tests/protractor_desktop/explorationHistoryTab.js:31:1)
>         at Module._compile (internal/modules/cjs/loader.js:1063:30)
>         at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
>         at Module.load (internal/modules/cjs/loader.js:928:32)
>         at Function.Module._load (internal/modules/cjs/loader.js:769:14)
> ```
> 
> It doesn't look like one of the flakes we've seen before. Could you dig into it a little? Here's the build link: https://github.com/oppia/oppia/pull/12699/checks?check_run_id=2525204687. If you need help debugging, check out the E2E test wiki page at https://github.com/oppia/oppia/wiki/End-to-End-Tests

@U8NWXD, The error is coming from the action.getText() in line no : **688**. The error is because the element **lineNumberElement**  is taking too long to appear. I don't know how to fix it. I guess the fault is not with the code I have changed. Please help me. TX.

U8NWXD commented:
@suryasiriki4 I think this is because of the changes in this PR. Here is the diff around the line that's throwing the error:

```diff
- var lineNumber = await lineNumberElement.getText();
+ var lineNumber = await action.getText(
+   'Line number element', lineNumberElement);
```

The error is being thrown because `action.getText` waits for the element to be visible, while before we called `.getText()` without waiting. I suspect that in some cases `lineNumberElement` is not visible. In these cases, `lineNumberElement.getText()` might return `''` instead of throwing an error. You could check this by adding some debugging statements to print out whether the element is visible and the result of calling `.getText()` on it. For example:

```js
console.log(`Visibility: ${lineNumberElement.isDisplayed()}`)
console.log(`Text: ${lineNumberElement.getText()}`)
console.log(`Visibility: ${lineNumberElement.isVisible()}`)
var lineNumber = await lineNumberElement.getText();
```

If my suspicion is correct, we would expect to see output like this:

```text
...
Visibility: False
Text:
Visibility: False
...
```

Here's the documentation for `isDisplayed()`: https://www.protractortest.org/#/api?view=webdriver.WebElement.prototype.isDisplayed

I didn't realize that `getText()` behaves this way, so I'm looking forward to seeing what you find!

suryasiriki4 committed:
Merge branch 'develop' into fix-forms
